### PR TITLE
[ML] validate remote cluster version when datafeed contains runtime_mappings

### LINKED ISSUE #69187: [ML] Better handle cross cluster search failures due to runtime mappings
When using runtime_mappings in a datafeed and executing a cross cluster search that hits an older cluster, the datafeed only writes out an audit message indicating the search failed.

It would be better to prevent the datafeed from being started at all. Indicating that one or more of the remote clusters don't support runtime_mappings.

Other things like new aggregations, etc. are more complicated. But, since runtime_mappings are a new field in the datafeed, it would be nice to indicate if they aren't supported by all the clusters in the source indices.

Example:

```
PUT _ml/anomaly_detectors/flight-count
{
  "description" : "count",
  "analysis_config" : {
    "bucket_span":"15m",
    "detectors": [
      {
        "detector_description": "count",
        "by_field_name": "day_of_week",
        "function": "count"
      }
    ]
  },
  "data_description" : {
    "time_field":"timestamp",
    "time_format": "epoch_ms"
  }
}
```

```
PUT _ml/datafeeds/datafeed-flight-count
{
  "job_id": "flight-count",
  "indices": ["cluster_zero:kibana*"],
  "runtime_mappings": {
  "day_of_week": {
    "type": "keyword",
    "script": {
      "source": "emit(doc['timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
    }
  }
}
}
```

This results in an audit message:
```
"message": "Datafeed is encountering errors extracting data: Versions before 7.11.0 don't support [runtime_mappings] and search was sent to [7.9.4]",
                    "level": "error",
```

It would be nice to prevent starting the datafeed altogether.

### PR DESCRIPTION
**Pull request by benwtrent**
Recently `runtime_mappings` were added to the datafeed config.

Since datafeeds support cross-cluster-search, we should validate the remote cluster connection versions.

We only need to do this when runtime_mappings are defined in the datafeed config.

Previously, the search would just fail when running against the remote cluster and write an error audit message.

close https://github.com/elastic/elasticsearch/issues/69187

### CODE REVIEW

**Comment by benwtrent** [comment]
This is what the error looks like (ran manual tests)
```
{
    "error": {
        "root_cause": [
            {
                "type": "status_exception",
                "reason": "remote clusters [other_cluster] are not at least version [7.11.0] which is required for [runtime mappings]"
            }
        ],
        "type": "status_exception",
        "reason": "cannot start datafeed [simple_count-datafeed] as it uses indices on a remote cluster [other_cluster:kibana*] but the license type could not be verified",
        "caused_by": {
            "type": "status_exception",
            "reason": "remote clusters [other_cluster] are not at least version [7.11.0] which is required for [runtime mappings]"
        }
    },
    "status": 400
}
```

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM Just a suggestion for reusing the definition of `runtime_mappings`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java [hunk 1]
```diff
@@ -286,6 +291,12 @@ public Integer getScrollSize() {
         return scrollSize;
     }
 
+    public Optional<Tuple<Version, String>> minRequiredClusterVersion() {
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:294]: Just FYI: In my PR introducing remote cluster version validation, I decided to replace `TransformConfig.minRequiredClusterVersion` with `TransformConfig.getAdditionalValidations` the reason being code duplication.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedActionTests.java [hunk 2]
```diff
@@ -101,6 +108,54 @@ public void testNoDeprecationsLogged() {
...
     }
 
+    public void testRemoteClusterVersionCheck() {
+        Map<String, Version> clusterVersions = MapBuilder.<String, Version>newMapBuilder()
+            .put("modern_cluster_1", Version.CURRENT)
+            .put("modern_cluster_2", Version.CURRENT)
+            .put("old_cluster_1", Version.V_7_0_0)
+            .map();
+
+        Map<String, Object> settings = new HashMap<>();
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedActionTests.java:118]: Could `MapBuilder` be used here too?

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedActionTests.java [hunk 3]
```diff
@@ -101,6 +108,54 @@ public void testNoDeprecationsLogged() {
...
+        settings.put("type", "keyword");
+        settings.put("script", "");
+        Map<String, Object> field = new HashMap<>();
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedActionTests.java:121]: Could `singletonMap` or `MapBuilder` be used here?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java [hunk 4]
```diff
@@ -243,6 +260,31 @@ public void onFailure(Exception e) {
...
+
+        List<String> clustersTooOld = remoteClusters.stream()
+            .filter(cn -> clusterVersionSupplier.apply(cn).before(minVersion))
+            .collect(Collectors.toList());
+        if (clustersTooOld.isEmpty()) {
+            return;
+        }
+
+        throw ExceptionsHelper.badRequestException(
+            "remote clusters {} are not at least version [{}] which is required for [{}]",
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java:281]: Please take a look at the analogical (but different) message in `SourceDestValidator.REMOTE_CLUSTERS_TOO_OLD` and try to unify them.

**Review by przemekwitek — APPROVED** [review]
LGTM

**Comment by benwtrent** [comment]
Note: For datafeeds, the only place we call for field caps is in the scroll extractor. And only then do we really care about the timefield. 

Since we block using runtime_mappings for the time field, we are filtering out any of the runtime_fields from the capabilities request and are NOT setting the `runtime_fields` option. 

So, the runtime mappings check should be for 7.11 and not for 7.12. Since we don't use the field caps API in conjunction with runtime_mappings (which was not introduced until 7.12)

**Pull request merged**